client/src/instance_spec.rs

```diff
@@ -0,0 +1,287 @@
+    //   compatibility themselves, which is an affordance the migration protocol
+    //   allows).
+    pub json: String,
+}
+
+impl TryFrom<&CrucibleRequestContents> for crucible::VolumeConstructionRequest {
+    type Error = serde_json::Error;
+
+    fn try_from(value: &CrucibleRequestContents) -> Result<Self, Self::Error> {
+        serde_json::from_str(&value.json)
+    }
+}
+
+impl TryFrom<&crucible::VolumeConstructionRequest> for CrucibleRequestContents {
+    type Error = serde_json::Error;
+
+    fn try_from(
+        value: &crucible::VolumeConstructionRequest,
+    ) -> Result<Self, Self::Error> {
+        Ok(Self { json: serde_json::to_string(value)? })
+    }
+}
+
+/// A kind of storage backend: a connection to on-sled resources or other
+/// services that provide the functions storage devices need to implement their
+/// contracts.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub enum StorageBackendKind {
+    /// A Crucible-backed device, containing a generation number and a
+    /// construction request.
+    Crucible { gen: u64, req: CrucibleRequestContents },
+
+    /// A device backed by a file on the host machine. The payload is a path to
+    /// this file.
+    File { path: String },
+
+    /// A device backed by an in-memory buffer in the VMM process. The payload
+    /// is the buffer's contents.
+    InMemory { bytes: Vec<u8> },
+}
+
+/// A storage backend.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct StorageBackend {
+    /// The kind of storage backend this is.
+    pub kind: StorageBackendKind,
+
+    /// Whether the storage is read-only.
+    pub readonly: bool,
+}
+
+/// A kind of storage device: the sort of virtual device interface the VMM
+/// exposes to guest software.
+#[derive(Clone, Copy, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub enum StorageDeviceKind {
+    Virtio,
+    Nvme,
+}
+
+/// A storage device.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct StorageDevice {
+    /// The device interface to present to the guest.
+    pub kind: StorageDeviceKind,
+
+    /// The name of the device's backend.
+    pub backend_name: String,
+
+    /// The PCI path at which to attach this device.
+    pub pci_path: PciPath,
+}
+
+//
+// Network devices.
+//
+// Because there is only one kind of virtual networking device (a virtio-net
+// device with an enlightened vNIC), networking devices don't have the same
+// BackendKind and DeviceKind enums that storage devices do, but they can
+// be added if needed in future versions.
+
+/// A network backend, specifically a virtual NIC on the host system for which
+/// Virtio network adapter integration is enabled.
+#[derive(Clone, Deserialize, Serialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct NetworkBackend {
+    /// The name of the vNIC to connect to.
+    pub vnic_name: String,
```

Comment:
Do we want to include a `kind` here, in case we expand beyond viona at some point?